stl/inc/stop_token

```diff
@@ -0,0 +1,401 @@
+    // recheck the state in case it changed while we were waiting to acquire the lock
+    _Local_sources = _State->_Stop_sources.load();
+    if ((_Local_sources & uint32_t{1}) != 0) {
+        // stop already requested
+        _State->_Callbacks._Store_and_unlock(_Head);
+        _Fn(this);
+        return;
+    }
+
+    if (_Local_sources != 0) {
+        // stop possible, do the insert
+        _Parent = _State;
+        _Next   = _Head;
+        if constexpr (_Transfer_ownership) {
+            _State_raw = nullptr;
+        } else {
+            _State->_Stop_tokens.fetch_add(1, memory_order_relaxed);
+        }
+
+        if (_Head != nullptr) {
+            _Head->_Prev = this;
+        }
+
+        _Head = this;
+    }
+
+    _State->_Callbacks._Store_and_unlock(_Head);
+}
+
+inline void _Stop_callback_base::_Attach(const stop_token& _Token) noexcept {
+    this->_Do_attach<false>(_Token._State);
+}
+
+inline void _Stop_callback_base::_Attach(stop_token&& _Token) noexcept {
+    this->_Do_attach<true>(_Token._State);
+}
+
+inline void _Stop_callback_base::_Detach() noexcept {
+    stop_token _Token{_Parent}; // transfers ownership
+    if (_Token._State == nullptr) {
+        // callback was never inserted into the list
+        return;
+    }
+
+    auto _Head = _Token._State->_Callbacks._Lock_and_load();
+    if (this == _Head) {
+        // we are still in the list, so the callback is not being request_stop'd
+        const auto _Local_next = _Next;
+        if (_Local_next != nullptr) {
+            _Local_next->_Prev = nullptr;
+        }
+
+        _STL_INTERNAL_CHECK(_Prev == nullptr);
+        _Token._State->_Callbacks._Store_and_unlock(_Next);
+        return;
+    }
+
+    const auto _Local_prev = _Prev;
+    if (_Local_prev != nullptr) {
+        // we are still in the list, so the callback is not being request_stop'd, and there is at least one other
+        // callback still registered
+        const auto _Local_next = _Next;
+        if (_Local_next != nullptr) {
+            _Next->_Prev = _Local_prev;
+        }
+
+        _Prev->_Next = _Local_next;
+        _Token._State->_Callbacks._Store_and_unlock(_Head);
+        return;
+    }
+
+    // we aren't in the callback list even though we were added to it, so the stop requesting thread is attempting to
+    // call the callback
+    _STL_INTERNAL_CHECK((_Token._State->_Stop_sources.load() & uint32_t{1}) != 0);
+    if (_Token._State->_Current_callback.load(memory_order_relaxed) != this
+        || _Token._State->_Stopping_thread == _Thrd_id()) {
+        // the callback is done or the dtor is being recursively reentered, do not block
+        _Token._State->_Callbacks._Store_and_unlock(_Head);
+        return;
+    }
+
+    // the callback is being executed by another thread, block until it is complete
+    _Token._State->_Callbacks._Store_and_unlock(_Head);
+    _Token._State->_Current_callback.wait(this, memory_order_relaxed);

```

Comment:
I think this needs to be `memory_order_acquire` and the store to `_Curent_callback` in `_Request_stop` needs to use `memory_order_release` so that we establish a "happens before" relationship between the return from the callback and the destruction of the callback object  subsequent destruction of the callback object as required by `[stopcallback.cons]/6`.

Reply:
`_Lock_and_load` / `_Store_and_unlock` pair is a mutex, doesn't it help?

Reply:
At the time the `_Store_and_unlock()` call is made here the callback may not have finished executing on the other thread (this is why we need the subsequent `wait()`) so it won't establish a happens-before relationship with the return of the callback.

Reply:
Makes sense, thanks!